pkg/client/rootd/dns/connpool.go

```diff
@@ -0,0 +1,111 @@
+package dns
+
+import (
+	"container/heap"
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"time"
+
+	"github.com/miekg/dns"
+)
+
+type ConnPool struct {
+	items      map[*dns.Conn]bool
+	newArrival chan *waitingClient
+	finished   chan *dns.Conn
+	clients    clientQueue
+	cancel     context.CancelFunc
+	RemoteAddr string
+}
+
+func NewConnPool(addr string, poolSize int) *ConnPool {
+	cCtx, cCancel := context.WithCancel(context.Background())
+	pool := &ConnPool{
+		items:      make(map[*dns.Conn]bool, poolSize),
+		newArrival: make(chan *waitingClient),
+		finished:   make(chan *dns.Conn),
+		cancel:     cCancel,
+		RemoteAddr: addr,
+	}
+	heap.Init(&pool.clients)
+	for i := 0; i < poolSize; i++ {
+		conn, err := dns.Dial("udp", net.JoinHostPort(addr, "53"))
+		if err != nil {
+			fmt.Printf("failed to create conn: %v", err)
+			os.Exit(1)
```

Comment:
I second this. We cannot just exit since that'll bypass all deferred functions everywhere. A `runtime.Goexit()` could be feasible in some cases but a simple `panic` with the error, accompanied by a recover that takes care of the error where the goroutine is started is almost always better in cases where the error cannot be propagated by just returning it.

Reply:
oops this is a copy paste error, will fix